pandas/_libs/parsers.pyx

```diff
@@ -1278,8 +1280,18 @@ cdef class TextReader:
                 indexer = cats.get_indexer(unsorted)
                 codes = take_1d(indexer, codes, fill_value=-1)
 
-            return Categorical(codes, categories=cats, ordered=False,
-                               fastpath=True), na_count
+            cat = Categorical(codes, categories=cats, ordered=False,
+                              fastpath=True)
+
+            if isinstance(dtype, CategoricalDtype):
+                if dtype.categories is None:
+                    # skip recoding
+                    if dtype.ordered:
+                        cat = cat.set_ordered(ordered=dtype.ordered)
+                else:
+                    cat = cat.set_categories(dtype.categories,
+                                             ordered=dtype.ordered)
```

Comment:
I wonder if you could refactor this a little and write it as such:
~~~python
if isinstance(dtype, CategoricalDtype):
   if dtype.categories is not None:
      cat = cat.set_categories(dtype.categories)
   cat = cat.set_ordered(ordered=dtype.ordered)
~~~